### PR TITLE
components only display its relevant feature flags

### LINKED ISSUE #46822: Make "--feature-gates" flag component-related
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/.):

**Note:** Please file issues for subcomponents under the appropriate repo

| Component | Repo                                                               |
| --------- | ------------------------------------------------------------------ |
| kubectl   | [kubernetes/kubectl](https://github.com/kubernetes/kubectl/issues/new) |
| kubeadm   | [kubernetes/kubeadm](https://github.com/kubernetes/kubeadm/issues/new) |

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Enhancement.

**Kubernetes version** (use `kubectl version`):
HEAD

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
We have `--feature-gates` flag in help info of every main components. And all of them have the same options listed.

```
Accelerators=true|false (ALPHA - default=false)
AdvancedAuditing=true|false (ALPHA - default=false)
AffinityInAnnotations=true|false (ALPHA - default=false)
AllAlpha=true|false (ALPHA - default=false)
AllowExtTrafficLocalEndpoints=true|false (default=true)
AppArmor=true|false (BETA - default=true)
DynamicKubeletConfig=true|false (ALPHA - default=false)
DynamicVolumeProvisioning=true|false (ALPHA - default=true)
ExperimentalCriticalPodAnnotation=true|false (ALPHA - default=false)
ExperimentalHostUserNamespaceDefaulting=true|false (BETA - default=false)
PersistentLocalVolumes=true|false (ALPHA - default=false)
StreamingProxyRedirects=true|false (BETA - default=true)
TaintBasedEvictions=true|false (ALPHA - default=false)
```

When users want to switch a feature gate on, it's hard to know which components need to enable the flag. Take `AffinityInAnnotations` for example, users might know they should enable this flag for apiserver and scheduler. But how about controller-manager and kubelet? They might need or not. For kube-proxy, it obviously doesn't need to enable this flag.

**What you expected to happen**:
Only show the options for components which need them. It would be good to have a documentation to guide users.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

/sig api-machinery

### PR DESCRIPTION
**Pull request by dqminh**
**What this PR does / why we need it**:
Currently kubernetes components display all features in `--help` text even if the
features do not apply to a specific component. This change adds
`Components` to `FeatureSpec` and use that to manipulate the `--help`
text to display only relevant feature flag to a component.

For example, kubelet features gate before:

```
--feature-gates mapStringBool                              A set of key=value pairs that describe feature gates for alpha/experimental features. Options are:
APIListChunking=true|false (BETA - default=true)
APIResponseCompression=true|false (ALPHA - default=false)
Accelerators=true|false (ALPHA - default=false)
AdvancedAuditing=true|false (BETA - default=true)
AllAlpha=true|false (ALPHA - default=false)
AllowExtTrafficLocalEndpoints=true|false (default=true)
AppArmor=true|false (BETA - default=true)
CPUManager=true|false (ALPHA - default=false)
CustomResourceValidation=true|false (ALPHA - default=false)
DebugContainers=true|false (ALPHA - default=false)
DevicePlugins=true|false (ALPHA - default=false)
DynamicKubeletConfig=true|false (ALPHA - default=false)
EnableEquivalenceClassCache=true|false (ALPHA - default=false)
ExpandPersistentVolumes=true|false (ALPHA - default=false)
ExperimentalCriticalPodAnnotation=true|false (ALPHA - default=false)
ExperimentalHostUserNamespaceDefaulting=true|false (BETA - default=false)
HugePages=true|false (ALPHA - default=false)
Initializers=true|false (ALPHA - default=false)
KubeletConfigFile=true|false (ALPHA - default=false)
LocalStorageCapacityIsolation=true|false (ALPHA - default=false)
MountPropagation=true|false (ALPHA - default=false)
PersistentLocalVolumes=true|false (ALPHA - default=false)
PodPriority=true|false (ALPHA - default=false)
RotateKubeletClientCertificate=true|false (BETA - default=true)
RotateKubeletServerCertificate=true|false (ALPHA - default=false)
ServiceNodeExclusion=true|false (ALPHA - default=false)
StreamingProxyRedirects=true|false (BETA - default=true)
SupportIPVSProxyMode=true|false (ALPHA - default=false)
TaintBasedEvictions=true|false (ALPHA - default=false)
TaintNodesByCondition=true|false (ALPHA - default=false)
```

kubelet features gate after this change:

```
--feature-gates mapStringBool                               A set of key=value pairs that describe feature gates for alpha/experimental features. Options are:
Accelerators=true|false (ALPHA - default=false)
AppArmor=true|false (BETA - default=true)
CPUManager=true|false (ALPHA - default=false)
DevicePlugins=true|false (ALPHA - default=false)
DynamicKubeletConfig=true|false (ALPHA - default=false)
ExperimentalCriticalPodAnnotation=true|false (ALPHA - default=false)
ExperimentalHostUserNamespaceDefaulting=true|false (BETA - default=false)
HugePages=true|false (ALPHA - default=false)
KubeletConfigFile=true|false (ALPHA - default=false)
LocalStorageCapacityIsolation=true|false (ALPHA - default=false)
MountPropagation=true|false (ALPHA - default=false)
PersistentLocalVolumes=true|false (ALPHA - default=false)
PodPriority=true|false (ALPHA - default=false)
RotateKubeletClientCertificate=true|false (BETA - default=true)
RotateKubeletServerCertificate=true|false (ALPHA - default=false)
```

Since this only changes `--help` text, any previous usage of invalid feature gate for a component
will still be accepted.

**Which issue(s) this PR fixes***:
Fix #46822

**Special notes for your reviewer**:
List of components for each features was derived manually from function usages, package imports etc. Its very much possible that they can be wrong.

cc @luxas @liggitt @xiangpengzhao

### CODE REVIEW

**Comment by xiangpengzhao** [comment]
I like this PR.
/cc @jamiehannaford 
who had been working on the similar things.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go
Comment (xiangpengzhao) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go]: Can here reuse the const defined in `pkg/features/kube_features.go` ?

File: pkg/features/kube_features.go
Comment (xiangpengzhao) [inline pkg/features/kube_features.go]: Can you make sure all the features are exactly associated to the components you list here ?

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (tengqm) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: We have got many questions related to this mapping. Many many thanks for figuring this out.

One improvement I'd like to suggest is to introduce 3 additional string fields:

```
  // IntroducedSince is the release when a feature is first introduced
  IntroducedSince: string
  // DeprecatedSince is the release when the feature is marked as deprecated
  DeprecatedSince: string
  // RemovedSince is the release when a feature is removed
  RemovedSince: string
```

The `RemovedSince` field can be ignored because it disappears anyway, however the other two fields would be very helpful for users.

Comment (xiangpengzhao) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Sounds like a good idea. However, we have `PreRelease: utilfeature.Alpha, utilfeature.Beta, utilfeature.GA`, how can these versions be mapping to your suggested fields?

File: pkg/features/kube_features.go
Comment (jamiehannaford) [inline pkg/features/kube_features.go]: while we're at it, can we improve the sorting here so that it's either alphabetical or grouped by component?

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go]: is it worth defining a new method to `k8s.io/apiserver/pkg/util` that's something like `AddOrPanic`? we're redefining the functionality a few times here

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: I like this idea too. I think `IntroducedSince` could also be a `Version` struct, since that provides more semantic context than a string (which can be malformed). IMO the `PreRelease` is something slightly orthogonal to what version a feature was added.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: what about just `reflect.DeepEquals`?

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: what about:

```
_, ok := f.Components[component]
return ok
```

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: since we're effectively using Enums here, how about defining a new `type Component string` alias, then specifying it here? Prevents people getting it wrong then

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: to avoid iterating twice and using 2 structs, you can do

```
for k, v := range allFeatures {
   if !v.hasComponent(component) {
      continue
   }
   ...
```

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go]: can we also test the new functionality added by the second arg?

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go
Comment (jamiehannaford) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go]: we also don't seem to be testing new functionality for `KnownFeatures`

**Review by jamiehannaford — CHANGES_REQUESTED** [review]
this is great! thanks for submitting this. here's some first feedback

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Components is only a slice though, not a set so we cant use this. I think a slice is ok because its size is very small.

Comment (jamiehannaford) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Ah okay

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go]: I think this is being tested with the new KnownFeatures test now so i dont think its required to modify the current test.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate_test.go]: Thanks. Added.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Maybe this should be another change ? As of now i dont think the added metadata ( that replaces/augments current versioning metadata in feature comments ) makes much sense with the current implementation since they shouldnt appear in flag text.

This probably makes more sense if we autogenerate some form of documentation similar to the matrix in https://github.com/kubernetes/website/issues/6141#issuecomment-341426400 for example.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Fixed.

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/features/kube_features.go]: Added `MustAddDefault` for this.

File: pkg/features/kube_features.go
Comment (dqminh) [reply pkg/features/kube_features.go]: Sorted alphabetically now.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: added Component type.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Fixed.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (jamiehannaford) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Adding these fields would serve more than just docs, it would allow us to, for example, validate that the feature can be used with a specific component version. But yeah, we can do this in a follow-up PR if you'd prefer.

**Review by jamiehannaford — APPROVED** [review]

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (dqminh) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: >  it would allow us to, for example, validate that the feature can be used with a specific component version

Indeed, but it probably should be done carefully. What should we do if the validation failed ? We can prevent the component from starting but that will probably break some installations if people adding their required flags to all components ( that's what i have been doing before this change :stuck_out_tongue_closed_eyes: )

Comment (jamiehannaford) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Yeah, good points. Would you mind creating an issue so we can track this feature conversation?

File: pkg/features/kube_features.go
Comment (dqminh) [reply pkg/features/kube_features.go]: Not sure how we can be certain of this, most likely all features maintainer will need to take a look ? The current changes was determined manually by functions and packages thats involved `utilfeature.DefaultFeatureGate.Enabled` check

Comment (liggitt) [reply pkg/features/kube_features.go]: That's my main concern with this change... gates are intended to be a global toggle of a feature that is often cross-cutting across multiple components. I'd expect our recommendation to be to set `--feature-gates` identically for all kube components, not try to worry about which gates touch aspects of which components.

Comment (dqminh) [reply pkg/features/kube_features.go]: @liggitt however setting feature-gates to be the same for all components seems a bit too heavy handed / not user-friendly, especially for cluster operator. Now the only way to be sure a feature is set is to set on all of them, which seems very wasteful ( requires cluster rolling restart even for not related components for example ).

**Comment by liggitt** [comment]
hoisting to the top-level:

> @liggitt:
> That's my main concern with this change... gates are intended to be a global toggle of a feature that is often cross-cutting across multiple components. I'd expect our recommendation to be to set --feature-gates identically for all kube components, not try to worry about which gates touch aspects of which components.

> @dqminh:
> however setting feature-gates to be the same for all components seems a bit too heavy handed / not user-friendly, especially for cluster operator. Now the only way to be sure a feature is set is to set on all of them, which seems very wasteful ( requires cluster rolling restart even for not related components for example ).

I agree that "the only way to be sure a feature is set is to set on all of them", but that is still the case after this PR... this PR just hides some feature gates based on non-authoritative lists of which components use which flags. It is not always clear which components make use of some library-like portions of the code base.

**Comment by dqminh** [comment]
> It is not always clear which components make use of some library-like portions of the code base.

Indeed, this is my concern too when writing this change `List of components for each features was derived manually from function usages, package imports etc. Its very much possible that they can be wrong`

I'm not sure how to resolve this cleanly / or at all yet with the current behavior. I was hoping that feature authors maybe could help identify the components features are intended to.

> "the only way to be sure a feature is set is to set on all of them", but that is still the case after this PR... 

This is also true, the current change only hides features when showing the flags. You can still set it via the flag even if the feature is "hidden" and its intended because i dont want to break existing setups which may sets features to all components.

**Comment by liggitt** [comment]
> because i dont want to break existing setups which may sets features to all components.

not only should we not break that approach, that would be my recommended approach

**Comment by luxas** [comment]
I do think it's annoying to not clearly know what gates need to be set on which components, but I also agree that recommending the approach to "set the gate everywhere" is the easiest thing to do here, as the dependency between the different features and clearing out which component actually needs the gate is non-obvious.

This is actually something we could bring up to @kubernetes/sig-architecture-feature-requests some time, but it's not urgent IMO. As we break out components from the core repo and break some of the hard linkages between them, this might be a thing worth looking at.

I see a couple of options for going forward here:
1) Close the PR for now as out of scope or "not the design we want atm"
2) Merge the PR in a way that preserves backwards-compat (feature gates exist for all component like before) but making it possible to add _new_ feature gates per-component if we create such features in the future; i.e. slowly migrating away from what we have now
3) Merge the PR, hashing out all dependencies & connections, removing unneeded gates in alpha state this release and keeping unneeded beta gates for one cycle (v1.9) (the most aggressive option)

WDYT?

**Comment by jamiehannaford** [comment]
My understanding is that this PR only changes the way flags are shown to the user. Features are still registered and consumed throughout the codebase in the same way (i.e. it's backward compatible). 

I can't see anywhere else where an authoritative list is defined, so why not make a precedent here?

**Comment by dqminh** [comment]
Yes, @jamiehannaford is correct. This only changes the way flags are shown to user so we have a better understanding of where to set the flags, hence not having to restart components unnecessarily with unnecessary changes.

So this is basically @luxas 2), but only adding a hint of where to set feature gate so its backward-compatible.

> This is actually something we could bring up to @kubernetes/sig-architecture-feature-requests some time, but it's not urgent IMO. As we break out components from the core repo and break some of the hard linkages between them, this might be a thing worth looking at.

Yes this is also great and should help the use cases here a lot. For example one question i have seen a lot when setting feature gates is how out of place kube-proxy feature gates looks like when our rule of thumb is to set everything the same ( because some features are obviously not kube-proxy related :stuck_out_tongue:  )

**Comment by jfoy** [comment]
> For example one question i have seen a lot when setting feature gates is how out of place kube-proxy feature gates look

For what it's worth, I found this PR while trying to answer exactly this question. This change would have prevented some confusion.

**Comment by bgrant0607** [comment]
@mtaufen Have feature gates come up in component config?

**Comment by mtaufen** [comment]
@bgrant0607 yes, feature gates are typically included in componentconfig types as a `map[string]bool` field.

The expectation is that all `--feature-gates` style flags will be removed (like all the other flags), in favor of componentconfig APIs. Since componentconfig API types are defined in `.go` files, using the standard API machinery, I'd recommend including the feature gates applicable to the component in the inline docs that accompany the type definitions for now. 

That said, when feature gates were invented, there was a general expectation that any given feature gate could be a cross-component gate. So any change that checks a gate for the first time in a given component would need to update that component's componentconfig docs as well. 

I don't think there's a standard mechanism for generating command-line help text based on the componentconfig APIs, but IIRC kube-proxy may have at least had something to spit out the default yaml. (@ncdc?).

**Comment by ncdc** [comment]
> I don't think there's a standard mechanism for generating command-line help text based on the componentconfig APIs

I have an old, open PR for this, but it required that I write my own yaml emitter. I'd be happy to continue the conversation on it and decide how to handle it. #52198

> IIRC kube-proxy may have at least had something to spit out the default yaml.

Yes, that's the `--write-config-to` flag.

**Comment by luxas** [comment]
periodic poke on finding some conclusion on which way to go for this

**Pull request closed without merging**
